bpo-18060: Updating _fields_ of a derived struct type yields a bad cif

websurfer5 opened the pull request:
Use the size of the _fields_ elements array in the parent class of a Srtructure or Union derived subclass when cloning or modifying the _fields_ array. The old behavior of using the number of elements in the parent class can cause the array to be truncated when it is copied, especially when there are multiple layers of subclassing, which can result in segfaults or exceptions. Tests are included.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-18060](https://bugs.python.org/issue18060) -->
https://bugs.python.org/issue18060
<!-- /issue-number -->

websurfer5 committed:
bpo-18060: PyCStgDict_clone() needs to account for the length of
the base class when computing the length of the derived class.

websurfer5 committed:
bpo-18060: use the size of the _fields_ elements array in the parent
class of a Srtructure or Union derived subclass when cloning or
modifying the _fields_ array. The old behavior of using the number
of elements in the parent class can cause the array to be truncated
when it is copied, especially when there are multiple layers of
subclassing. Tests are included.

websurfer5 committed:
bpo-18060: remove extraneous debug code

websurfer5 committed:
bpo-18060: disable tests on Win32 platforms

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

websurfer5 committed:
bpo-18060: enable win32 tests

websurfer5 committed:
bpo-18060: take alignment into account when calculating number of elements in parent class

websurfer5 committed:
Merge branch 'master' of github.com:python/cpython into fix-issue-18060

serhiy-storchaka committed:
Merge branch 'main' into fix-issue-18060

serhiy-storchaka committed:
Add a newline.

serhiy-storchaka commented:
I think there is a misunderstanding from your side. `size` is not the size of the `ffi_type_pointer.elements` array. It is the size of the structure or the union.

serhiy-storchaka committed:
Correctly determine the size of StgDictObject.ffi_type_pointer.elements.

Now StgDictObject.length of Structure and Union is the number of all
fields, including fields of superclases.

serhiy-storchaka commented:
New tests are failed on the following buildbots:
https://buildbot.python.org/all/#/builders/256/builds/1322
https://buildbot.python.org/all/#/builders/390/builds/1322
https://buildbot.python.org/all/#/builders/356/builds/1247
https://buildbot.python.org/all/#/builders/111/builds/1264
https://buildbot.python.org/all/#/builders/403/builds/1266
https://buildbot.python.org/all/#/builders/239/builds/1273
https://buildbot.python.org/all/#/builders/232/builds/1254
https://buildbot.python.org/all/#/builders/178/builds/1451
https://buildbot.python.org/all/#/builders/370/builds/1290
https://buildbot.python.org/all/#/builders/177/builds/1298
https://buildbot.python.org/all/#/builders/170/builds/1285
https://buildbot.python.org/all/#/builders/420/builds/1281
https://buildbot.python.org/all/#/builders/276/builds/1291
https://buildbot.python.org/all/#/builders/558/builds/1511
https://buildbot.python.org/all/#/builders/332/builds/1321

serhiy-storchaka committed:
Refactor tests and skip them on big-endian platforms.
